torch/csrc/jit/ir.h

```diff
@@ -60,6 +61,67 @@ static inline bool operator==(const Use & a, const Use & b) {
 // Graph holds a list of parameters.
 struct Param;
 
+// Scope is a node of a trie that represents the tree of nested scopes.
+// Individual scopes are pushed and popped from Graph, which hold a
+// pointer to the current scope. Each Node in Graph holds a pointer
+// to the scope that was current when the node was created.
+// The trie never needs to shrink, it only grows until it is disposed
+// of when Graph is deallocated. Hence, pointers to scopes held by nodes
+// will always be valid.
+// Note that the ownership model as specified in the destructor implies 
+// that you cannot share common subtries between tries.
+struct Scope {
+private:
+  Scope* parent_;
+  Symbol name_;
+  std::unordered_set<const Scope*> children_;
```

Comment:
So, Graph holds a pointer to a node in the trie (the current node) and another pointer to the root. They are both pointers to Scope. The current node is for sure a member of `children_` of some other node, yet I have to hold it as current.

I'm not sure how I would safely use `unique_ptr` to store pointers to children without having to  `get` the raw pointers out to set them as current. It seems it's still an unsafe thing to do.
Should we use a `shared_ptr` to avoid these troubles?

Reply:
It seems to me that using a `unique_ptr` for the root, and `Scope*` for current scope, while having `unique_ptr`s in the scope trie for children will be safe - current scope is guaranteed to stay alive, because it is present in the subtree rooted in `root_scope_`, so there exists a chain of `unique_ptr`s leading to that `Scope`. I think it's exactly the same ownership model as proposed in this PR.

Calling `.get()` is safe is you can ensure that the pointer you received will be alive as long as you'll be using the result, which is the case here.

Reply:
Got it, thanks!